### PR TITLE
Add online feature

### PR DESCRIPTION
**Pull request by naxingyu**
Add part of the online features.
The OnlineCmvn binding is only reserved, not working. The issue here is that the members in OnlineCmvnState are of type Matrix<double> which is not bound.
Other features work fine.
@danpovey @csukuangfj

### CODE REVIEW

**Comment by danpovey** [comment]
You sure type Matrix is not bound?  how about this (from grepping around):
```
matrix/matrix_pybind.cc:  py::class_<Matrix<float>, MatrixBase<float>>(m, "FloatMatrix",
```

**Comment by danpovey** [comment]
BTW, might be easier if you make the PR to the corresponding branch in the Kaldi repo, we're moving this to there.

**Comment by naxingyu** [comment]
Matrix is only bound with float type, not double.

**Comment by naxingyu** [comment]
Yea I just saw that. I'll move it to the Kaldi repo once moved.

File: src/pybind/feat/online_feature_pybind.cc [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      "OnlineFeatureInterface, is not specified in the interface and you will"
+      "likely define new constructors or methods in the derived type to do that.")
+      .def("Dim", &OnlineFeatureInterface::Dim)
+      .def("NumFramesReady", &OnlineFeatureInterface::NumFramesReady)
+      .def("IsLastFrame", &OnlineFeatureInterface::IsLastFrame)
+      .def("GetFrame", &OnlineFeatureInterface::GetFrame)
+      .def("GetFrames", &OnlineFeatureInterface::GetFrames)
+      .def("FrameShiftInSeconds", &OnlineFeatureInterface::FrameShiftInSeconds);
+
+  py::class_<OnlineBaseFeature, OnlineFeatureInterface>(m, "OnlineBaseFeature");
```
Comment (csukuangfj) [inline src/pybind/feat/online_feature_pybind.cc]: two abstract methods are missing:

- AcceptWaveForms
- InputFinished

File: src/pybind/feat/online_feature_pybind.cc [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#include "feat/feature-fbank.h"
+#include "feat/online-feature.h"
+
+using namespace kaldi;
+  
+template <class Feature>
+void online_base_feature(py::module& m, const std::string& feat_type) {
+  py::class_<OnlineGenericBaseFeature<Feature>, OnlineBaseFeature>(m, feat_type.c_str())
+      .def(py::init<const typename Feature::Options&>())
+      .def("Dim", &OnlineGenericBaseFeature<Feature>::Dim)
```
Comment (csukuangfj) [inline src/pybind/feat/online_feature_pybind.cc]: I think `Dim` and other methods do not need to be re-wrapped here.
They are already wrapped in their parent classes, i.e., OnlineFeatureInterface.

File: src/pybind/feat/online_feature_pybind.cc [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  online_base_feature<MfccComputer>(m, "OnlineMfcc");
+  online_base_feature<FbankComputer>(m, "OnlineFbank");
+
+  py::class_<OnlineCmvnOptions>(m, "OnlineCmvnOptions")
+      .def(py::init<>())
+      .def_readwrite("cmn_window", &OnlineCmvnOptions::cmn_window)
+      .def_readwrite("speaker_frames", &OnlineCmvnOptions::speaker_frames)
+      .def_readwrite("global_frames", &OnlineCmvnOptions::global_frames)
+      .def_readwrite("normalize_mean", &OnlineCmvnOptions::normalize_mean)
```
Comment (csukuangfj) [inline src/pybind/feat/online_feature_pybind.cc:65]: `normalize_variance` is missing.

File: src/pybind/feat/online_feature_pybind.cc [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      .def_readwrite("skip_dims", &OnlineCmvnOptions::skip_dims)
+      .def("Check", &OnlineCmvnOptions::Check);
+  
+  py::class_<OnlineSpliceOptions>(m, "OnlineSpliceOptions")
+      .def(py::init<>())
+      .def_readwrite("left_context", &OnlineSpliceOptions::left_context)
+      .def_readwrite("right_context", &OnlineSpliceOptions::right_context);
+
+  py::class_<OnlineCmvn, OnlineFeatureInterface>(m, "OnlineCmvn")
+      .def(py::init<const OnlineCmvnOptions&, OnlineFeatureInterface*>())
```
Comment (csukuangfj) [inline src/pybind/feat/online_feature_pybind.cc]: , py::arg("opts"), py::arg("src")

**Comment by csukuangfj** [comment]
Do we need `Matrix<double>` ?

I guess current kaldi pybind supports `BaseFloat == float` only and `double` is not wrapped.

**Comment by csukuangfj** [comment]
By the way, DLPack supports only 32-bit single precision float. (https://github.com/dmlc/dlpack/blob/master/include/dlpack/dlpack.h#L89)

**Comment by csukuangfj** [comment]
https://github.com/danpovey/kaldi/blob/c61b1a54785dc2d285aec3b46863d225ef1fa39f/src/feat/online-feature.h#L269-L274

I think what @naxingyu means is that `Matrix<double>` is not wrapped.

**Comment by naxingyu** [comment]
not sure, double Matrix is rarely used. We should let @danpovey decide whether to wrap it or not. It does not have to interact with dlpack, if it is only used in feature pipeline. Also, I’m not sure if it is possible to change the OnlineCmvnState members to float Matrix.

**Comment by danpovey** [comment]
The OnlineCmvnState was made double out of concerns about roundoff.

Does the matrix not being wrapped affect the usability of OnlineCmvnState,
e.g. prevent it from being wrapped entirely?
Because it's not something the user would be likely to interact with.

**Comment by naxingyu** [comment]
@danpovey I can wrap it entirely without `Maxtirx<double>`, but all the non-inherit interfaces of `OnlineCmvn` will have to be hidden, and also the constructor that takes `OnlineCmvnState` as an argument.

**Comment by danpovey** [comment]
OK, maybe better to wrap Matrix<double> then.  Can just be as the other one
but without the buffer stuff.

File: src/pybind/feat/online_feature_pybind.cc (hunk 4 above)
Comment (naxingyu) [reply src/pybind/feat/online_feature_pybind.cc]: all constructors with arguments are updated

**Comment by danpovey** [comment]
Thanks.  So LMK if  you think this is ready to merge.  Please make a PR on the kaldi github though, we're moving things there.  You can put a link in a comment there to this one so we don't lose the context.

**Comment by naxingyu** [comment]
OK

**Comment by naxingyu** [comment]
@danpovey I bound `Matrix<double>` and add a copy constrcutor from `Matrix<float>` such that online CMVN can be initialized using a global state. Document is added to explain why this is not bound against numpy.
Once @csukuangfj approves, I'll make PR on Kaldi repo.

**Comment by danpovey** [comment]
Great, thanks!

**Comment by csukuangfj** [comment]
+2

**Pull request closed without merging**